packages/core/mesh/teleport/src/testing/test-extension-with-streams.ts

```diff
@@ -0,0 +1,244 @@
+//
+// Copyright 2022 DXOS.org
+//
+
+import assert from 'node:assert';
+import { randomBytes } from 'node:crypto';
+import { Duplex } from 'node:stream';
+
+import { Trigger } from '@dxos/async';
+import { log } from '@dxos/log';
+import { schema } from '@dxos/protocols';
+import { TestServiceWithStreams } from '@dxos/protocols/proto/example/testing/rpc';
+import { createProtoRpcPeer, ProtoRpcPeer } from '@dxos/rpc';
+
+import { ExtensionContext, TeleportExtension } from '../teleport';
+
+interface TestExtensionWithStreamsCallbacks {
+  onOpen?: () => Promise<void>;
+  onClose?: () => Promise<void>;
+}
+
+export class TestExtensionWithStreams implements TeleportExtension {
+  public readonly open = new Trigger();
+  public readonly closed = new Trigger();
+
+  public extensionContext: ExtensionContext | undefined;
+  private _rpc!: ProtoRpcPeer<{ TestServiceWithStreams: TestServiceWithStreams }>;
+
+  private readonly _streams = new Map<string, TestStream>();
+
+  constructor(public readonly callbacks: TestExtensionWithStreamsCallbacks = {}) {}
+
+  get remotePeerId() {
+    return this.extensionContext?.remotePeerId;
+  }
+
+  private async _loadStream(streamTag: string, interval = 5, chunkSize = 2048) {
+    assert(!this._streams.has(streamTag), `Stream ${streamTag} already exists.`);
+
+    const networkStream = await this.extensionContext!.createStream(streamTag, {
+      contentType: 'application/x-test-stream',
+    });
+
+    const streamEntry: TestStream = {
+      networkStream,
+      bytesSent: 0,
+      bytesReceived: 0,
+      sendErrors: 0,
+      receiveErrors: 0,
+      startTimestamp: Date.now(),
+    };
+
+    const pushChunk = () => {
+      streamEntry.timer = setTimeout(() => {
+        const chunk = randomBytes(chunkSize);
+
+        if (
+          !networkStream.write(chunk, 'binary', (err) => {
+            if (!err) {
+              streamEntry.bytesSent += chunk.length;
+            } else {
+              streamEntry.sendErrors += 1;
+            }
+          })
+        ) {
+          networkStream.once('drain', pushChunk);
+        } else {
+          process.nextTick(pushChunk);
+        }
+      }, interval);
+    };
+
+    pushChunk();
+
+    this._streams.set(streamTag, streamEntry);
+
+    networkStream.on('data', (data) => {
+      streamEntry.bytesReceived += data.length;
+    });
+
+    networkStream.on('error', (err) => {
+      streamEntry.receiveErrors += 1;
+    });
+  }
+
+  private _closeStream(streamTag: string): Stats {
+    assert(this._streams.has(streamTag), `Stream ${streamTag} does not exist.`);
```

Comment:
For all errors, logs, try to make in the form of "Stream does not exist: xxx" (which helps keep things uniform and helps with filtering logs).